### PR TITLE
chore(docker): Optimize dockerfile

### PR DESCRIPTION
**Pull request by mritd**
Optimize dockerfile

Signed-off-by: kovacs <mritd@linux.com>

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,10 +1,55 @@
-FROM swift:4.2.4
+# Build swift backend
+FROM swift:5.6 AS swift-builder
```
Comment (filipechagas) [inline Dockerfile:2]:
```suggestion
FROM --platform=linux/x86-64 swift:5.6 AS swift-builder
```

File: Dockerfile [hunk 2]
```diff
@@ -1,10 +1,55 @@
...
+RUN set -ex \
+    && apt update \
+    && apt install libssl-dev -y \
+    && swift package clean \
+    && swift package update \
+    && swift build -Xswiftc -static-stdlib -j $(nproc) -c release \
+    && mv $(swift build -Xswiftc -static-stdlib -c release --show-bin-path)/Run /tmp/Run
+
+# Build node frontend
+FROM node:lts AS node-builder
```
Comment (filipechagas) [inline Dockerfile:17]:
```suggestion
FROM --platform=linux/x86-64 node:lts AS node-builder
```

File: Dockerfile [hunk 3]
```diff
@@ -1,10 +1,55 @@
...
+
+WORKDIR /gitignore.io
+
+RUN set -ex \
+    && yarn install \
+    && yarn build \
+    && rm -rf node_models
+
+# Build final image
+FROM debian:stable-slim
```
Comment (filipechagas) [inline Dockerfile:29]:
```suggestion
FROM --platform=linux/x86-64 debian:stable-slim
```

File: Dockerfile [hunk 4]
```diff
@@ -1,10 +1,55 @@
...
 COPY . ./
-RUN git submodule update --init --recursive
-RUN swift package clean
-RUN swift build -c release
-RUN mkdir /app/bin
-RUN mv `swift build -c release --show-bin-path` /app/bin
-EXPOSE 8080
-ENTRYPOINT ["/app/bin/release/Run", "serve", "-e", "prod", "-b", "0.0.0.0"]
+COPY .git ./
+
```
Comment (filipechagas) [inline Dockerfile:43]:
```suggestion
RUN git submodule update --init --recursive

```

**Review by filipechagas — CHANGES_REQUESTED** [review]
Thank you for this excellent contribution!

I left a couple of comments for you. 
The `--platform=linux/x86-64` will make sure arm users can still build the image.
And the other one seems to be a typo.

Once these are taken care of, the PR is good to be merged 🙂

**Comment by mritd** [comment]
@filipechagas Can we use `swift:5.6-focal`?

image


I noticed that it supports both x86 and arm64; this way x86 and arm64 users can not care about the image architecture, and we can also build images with multi-arch support.

**Comment by mritd** [comment]
I've modified the Dockerfile so that it now supports both x86 and arm64 builds (due to the swift image not supporting other CPU architectures for now).

It is now possible to use the `docker buildx` command to build a multi-CPU architecture-supported image (compatible with the standard `docker build` command), I have pre-built a version: `mritd/gitignore.io`, this image can be used to verify this PR .

image

By the way, I'm not familiar with the swift language, and I don't know if there are other compilation parameters that can optimize the size of the binary compiled by swift; if there are better compilation parameters, I can further optimize the image size.

**Review by filipechagas — APPROVED** [review]
Excellent! I can confirm it builds on ARM with no issues.
Thank you @mritd !!

File: Dockerfile [hunk 5]
```diff
@@ -1,10 +1,58 @@
...
+COPY --from=node-builder /gitignore.io/Resources /app/Resources
+
+SHELL ["/bin/bash", "-c"]
+
+EXPOSE 8080/tcp
+
+# Add dump-init to ensure container can respond to exit signals
+ENTRYPOINT ["/usr/bin/dumb-init", "--"]
+
+CMD /app/Run serve -e prod -b 0.0.0.0
```
Comment (filipechagas) [inline Dockerfile:58]:
```suggestion
CMD ["/app/Run", "serve", "-e", "prod", "-b", "0.0.0.0"]
```

**Comment by mritd** [comment]
@filipechagas I re-edited the CMD command and added some comments so anyone can build their own docker image.

**Review by filipechagas — CHANGES_REQUESTED** [review]
Thanks for applying the changes @mritd !

I just need to ask you for a final adjustment: The docker-compose commands aren't working anymore. We'll need to remove the `command` line from `docker-compose.yml`

As for `docker-compose-dev.yml` to work, there may be more work to do, probably using a `target`.

**Comment by mritd** [comment]
I noticed that there is a compose configuration created for development; I created a Dcokerfile (Dockerfile.dev) dedicated to development, but the vapor command fails to execute.

Since I'm not a swift developer, I don't know how to fix the vapor compilation problem (maybe I should downgrade the vapor toolbox version?), @filipechagas can you pass some sample commands to run the project via `vapor run` and some related version information?

This way I can recreate the dockerfile and compose files for development; I think the development environment should meet the following requirements:

- Map the current directory into the container to ensure native code runs
- Recompile swift code every time to ensure consistency
- Cache swift packages as much as possible to avoid repeated downloads

**Comment by filipechagas** [comment]
Sorry about the delay on this one. I was OOO.

I just realized that docker-compose for development wasn't really working already. So, at this point, I would apply your excellent changes in the `Dockerfile`. We can ditch `Dockerfie.dev` for now and work on that in a separate PR.

WDYT?

**Comment by mritd** [comment]
I think it's ok, do I need to undo the changes to Dockerfile.dev and docker-compose-dev.yaml?

**Comment by filipechagas** [comment]
Can you please? That way we avoid misleading people into thinking they're finished.

**Comment by mritd** [comment]
I have rolled back the changes to dev.

**Review by filipechagas — APPROVED** [review]
Thank you @mritd !

**Comment by mritd** [comment]
@filipechagas

**Comment by mritd** [comment]
No response for too long, I close this PR.

**Comment by konstrybakov** [comment]
Hey @mritd, very sorry for the long review! Is this still something you'd like to be merged? I'll have a quick look and move this forward

**Review by konstrybakov — APPROVED** [review]

**Review by rafilkmp3 — COMMENTED** [review]
LGTM

**Pull request merged**